### ISSUE TITLE
Feature request: nixos-assimilate:  Turn currently running system into NixOS

### ISSUE THREAD

**Issue by aszlig** [NixOS/nixpkgs#2079]
This command would not only be very useful for users to switch to NixOS but also would get [NixOps](https://github.com/NixOS/nixops) to support a large variety of different hosters. Also this would simplify the Hetzner bootstrapping process.

The implementation would look like this:
- Gather hardware, networking and maybe partition information.
- Evaluate configuration and build NixOS system.
- Inject a initrd+kernel somewhere in the current system (maybe search for a continuous region within the free disk space).
- Overwrite boot loader with pointers to the injected initrd+kernel.
- Sync
- `reboot -f`
- Erase disk and repartition (at least in the first implementation of `nixos-assimilate`, because detecting available partitions and reusing them could become quite difficult).
- Create filesystems and mount.
- Copy closures into the new filesystem layout.
- `pivot_root` into the new system and bootup NixOS.

In order to properly (re)partition, we need to do the refactoring of [nixpart](https://github.com/aszlig/nixpart) to first of all not be based on kickstart syntax and also have a sane way to specify filesystem layouts.

`nixpart` is based on [blivet](https://fedoraproject.org/wiki/Blivet), which works with partitioning information in a tree-like structure and also is able to detect existing filesystem layouts, so it's perfectly suited to work with Nix expressions.

That means we need to refactor `nixpart` to use Nix expressions to specify the filesystem layout and remove all kickstart-related stuff (kickstart is more than just a way to specify partitioning and has dependencies and complexity which we really don't need).

And of course, before we could do that, we need to come up with a sane layout for specifying partitions, possibly something that would blend in seamlessly with the `fileSystems` attribute set.

This possibly affects/fixes NixOS/nixops#168 and #1080.

**Comment by domenkozar** [comment]
@aszlig I'll bump this to 15.05, let me know if it's doable.

**Comment by CestDiego** [comment]
What is the status of this? Thank you guys for everything you do for the future of NixOps

**Comment by thoughtpolice** [comment]
Definitely not getting done for this release; punting

**Comment by aszlig** [comment]
Submitted a talk about it for NixCon 2015, so I'm going to work on this during October.

**Comment by jeaye** [comment]
This would keep me on DO, instead of forcing me onto Linode or some other providers. One key issue with that, however, is that DO makes some assumptions about how the filesystem is laid out. [I made an issue](https://github.com/gh2o/digitalocean-debian-to-arch/issues/7) on [digitalocean-debian-to-arch](https://github.com/gh2o/digitalocean-debian-to-arch) about this, which was ultimately worked out.

Either way, if other NixOS folks are trying this on DO, this will almost certainly become an issue.

**Comment by wmertens** [comment]
It seems to me that a 'light' version which throws away the current linux entirely and installs nixos with a desired configuration on the current partition layout is also desirable. Steps:

1. Get root access to system
2. Install temporary nixos under `/nixosconvert`. It should also get the desired configuration, which should fit the current partition layout and system.
3. Install nixos boot loader that bind mounts `/nixosconvert` to `/`
4. Reboot
5. Nixos converter system boots, mounts original root, clears it out except desired directories, and installs nixos on it. Boot loader gets overwritten with proper configuration
6. Reboot, welcome to your new NixOS.

I believe there was support for bind mounting in the initrd but it was taken out, not sure.

Other than that, this needs the nixos installer image (doesn't even need to be nixos itself) and a service that runs the install with the desired configuration, rebooting afterwards.

Thoughts?

**Comment by jeaye** [comment]
@wmertens I have developed such a "light" version, which is even simpler than that! Have a look at https://github.com/jeaye/nixos-in-place and my relevant post to the nix-dev list http://lists.science.uu.nl/pipermail/nix-dev/2015-December/018877.html

**Comment by wmertens** [comment]
@jeaye nice! In a perfect world it would install nixos without the ISO download, via a nix install which evaluates system, but this also works :)

**Comment by domenkozar** [comment]
@aszlig what's the status on this? Should we bump the milestone?

**Comment by elitak** [comment]
nixos-in-place wasn't working for me on Digital Ocean, so I made an even simpler version that you can pipe straight to bash over ssh, and uses the nix install script to bootstrap, rather than the ISO: https://gist.github.com/elitak/996e3b3c54c07409c8f2 . I'm next going to look into integrating the DO API into nixops to just instantiate a droplet and run my script to transform it to nixos.

I look forward to the final nixpart + dm-module solution, though; that should be pretty useful for installing on every platform.

**Comment by jeaye** [comment]
@elitak It's a shame to see yet another project trying to do this. It'd be great if you applied your spare time to working on nixos-assimilate or nixos-in-place instead!

**Comment by elitak** [comment]
@jeaye I hesitate to call it a "project". It's essentially a couple dozen lines of bash that I intend to discard once nixos-assimilate is ready. I don't consider it shameful to try to help everyone expedite their deployments by giving them something simple that works. I needed to write that script for my own use, whether I shared it or not.

I would like to contribute to nixos-assimilate, but it's pretty hard when none of it is shared. I can't find any work on the dm-module @aszlig mentioned in his nixcon talk, nor the nixpart changes. I checked aszlig/nixpkgs and aszlig/nixpart. If you know where the most recent work is published, please point me to it and I'd be more than happy to look and see what I can offer. Otherwise, I just have to wait until help is being accepted. I take the absence of updates here and in the repositories as an indication it isn't.

As for nixos-in-place, it didn't work as advertised to begin with (see notes in the comments of my script), so that route started with extra obstacles in it.

**Comment by aszlig** [comment]
Having `nixos-in-place` isn't really duplicate work, because it's working with the existing filesystem IIUC, while `nixos-assimilate` goal is to do repartitioning on the fly.

As for the published work, I'm currently focusing on `nixpart` and trying to rebase the nixpart branch on top of current master with the new upstream version(s). Unfortunately there are a few obstacles with the tests of `blivet` and `libblockdev`, because apparently it seems that not even the upstream project runs its own test suite :-/

I'll try to get the `nixpart` branch in nixpkgs finished over the next two weeks, so we should at least have the necessary dependencies ready to go with `nixpart`. The reason why I didn't yet push the new version of `nixpart` is that `blivet`s API was constantly changing at the time I started to work on it, so it's currently a bunch of experiments (part of those I have shown at the talk).

**Comment by elitak** [comment]
@aszlig I appreciate the update. I will have a cursory look at those test suites to see if there's anything easy to fix.

Where is the device-mapper module that will do the inode/block tagging? Is that part of libblockdev or is it still unwritten?

I still have a very nebulous picture of how it will all fit together. A repository tying them all together would be helpful. To that end, where can I find nixpart commits newer than 3 years ago?

If the answers to most of these questions are going to be self-evident in two weeks or so, I'm fine with waiting. I don't want to distract you from actual progress when its unlikely I'll be able to contribute without the groundwork you're doing anyway.

Thanks for your efforts on what I consider a real killer app for NixOS.

**Comment by 0xABAB** [comment]
@elitak `cat customConfig.optional nixos-infect | ssh root@targethost bash` is a fragile construct. If you make it such that you first copy the code to the host and then execute it in a remote `screen` for example, it wouldn't be. I didn't look further than that, but I am happy to see people moving in this direction.

**Comment by wmertens** [comment]
@elitak so you're saying that the changes introduced by https://github.com/NixOS/nixpkgs/commit/171d43ba4f51384670c8af0e5188c023f5bb0d94 are too brittle on other distros?

Can you perhaps try with `nixos-install` that does not have the above patch? Copying all those files around doesn't seem very efficient :(

**Comment by elitak** [comment]
I didn't realize gists didn't have issue tracking, so I've now promoted the gist to a project:
https://github.com/elitak/nixos-infect
Please file issues there so we keep comments here pertinent to nixos-assimilate.

@wmertens I only saw the "unshare" problem on Gentoo ARM hosts when I was performing the installation manually. The issue usually preventing bind/no-mount roots on cloud VPSes is the bash permission error one:
https://github.com/NixOS/nixpkgs/issues/10230

The installation needs to perform at least 2 copies anyway: one from the running system to the newroot, and one rsync back. The only cost incurred by that bug is that you need to zero out about 2GB for a loopback image. It's possible to do everything at once inside nixos-install, but that's out of the scope of my script. If I need dozens of NixOS VPSes, I will clone the first to the subsequent ones using either snapshots/imaging provided by the vendor or rsync -a --delete /tmp/rootbindmount/ secondhost:/tmp/rootbindmount/ (then reboot -f secondhost). I don't want to be putting that much load on the NixOS binary cache servers anyway; cloud-local rsync/snapshotting is the way to go.

@0xABAB, I know it's fragile; it's up to the user to know for himself if he needs something better. I'm just recommending the simplest, most optimistic way to do it. That optimism is a theme throughout the script. I want to keep it as simple as possible so that fixes can be easily made by users. The script is not idempotent, so the only thing screen really adds is avoidance of termination-by-disconnect.

nixos-assimilate, whenever it drops, will make the effort to handle every case. I will contribute to that endeavor as soon as I'm allowed to, starting with Digital Ocean support.

**Comment by wmertens** [comment]
@elitak why do you need to rsync back, why not simply move or bind mount root on a subdir?

**Comment by obadz** [comment]
I've got a PR for something related by less ambitious (no partition layout changes) if anyone is interested: #17784

**Comment by danbst** [comment]
@aszlig are there plans to publish the unfinished nixpart 1.0?

**Comment by aszlig** [comment]
@danbst: Not really, because I need to rewrite a lot of this stuff for latest blivet. But I could publish the unfinished packaging of blivet, because that's the biggest stopper right now and it would help to have it packaged (along with full NixOS tests).

**Comment by aszlig** [comment]
I just pushed the rebased `nixpart` branch along with working tests of `libblockdev`. Finishing the blivet tests ASAP (after sleep + work) and going to rewrite `nixpart` from scratch, because the `blivet` API changed a lot since the last time I was working on it. The upside is that the API has changed in a way that the actual implementation of `nixpart` should be only a matter of a few lines of code.

**Comment by aszlig** [comment]
See #21403.

**Comment by gilligan** [comment]
@aszlig i suppose this is still in progress somehow?

**Comment by roosemberth** [comment]
What's the state of this PR?

**Comment by danbst** [comment]
@gilligan @roosemberth 
I'll answer for @aszlig here, but he may correct me

This PR is blocked by https://github.com/NixOS/nixpkgs/pull/21403, without it being merged `nixos-assimilate` isn't much better than `nixos-infect`, `nixos-in-place` and co.

The `nixpart` PR itself is simply not finished. IMO it would be great to split `kickstart/blivet` part of it into separate PR and merge faster. It would then allow to kexec a running system into NixOS-in-RAM, do the partitioning (via kickstart) and install using normal way + some network initialization magic from `nixos-infect`. (https://nixos.wiki/wiki/Install_NixOS_on_a_Server_With_a_Different_Filesystem)

**Comment by petabyteboy** [comment]
I think it's not too hard to realize something like this using kexec. I am using @cleverca22's kexec files to automatically install NixOS to Hetzner Cloud VMs, though it is not integrated with NixOps. 
Basically I just took the kexec script and added a service that runs the necessary commands to install NixOS, and reboots.

**Comment by aszlig** [comment]
still important to me

**Comment by roosemberth** [comment]
Still important to me

**Comment by obadz** [comment]
Just want to advertise to people who might stumble on this issue that some (though not all) of the functionality that's aimed for here is available via [nixos-infect](https://github.com/elitak/nixos-infect) using [NIXOS_LUSTRATE](https://nixos.org/manual/nixos/stable/#sec-installing-from-other-distro).

**Comment by DrSensor** [comment]
.

**Comment by otavio** [comment]
This should get progress; it'd change how we can use NixOS in a number of cloud providers.

**Comment by mohe2015** [comment]
> This should get progress; it'd change how we can use NixOS in a number of cloud providers.

The official manual already contain LUSTRATE instructions that worked for me.

**Comment by obadz** [comment]
> This should get progress; it'd change how we can use NixOS in a number of cloud providers.

See also https://github.com/elitak/nixos-infect/

**Comment by nixinator** [comment]
it's 2022, probably good idea to use cloud providers that actively support nixos both thru supply bare metal images, or employ nix committers. Not every problem out there requires a technical hack or work around, sometimes is purely political pressure , the best way to exert pressure is through your $$$ cloud budget. We need a list of nixos friendly cloud providers! :-)

**Comment by mohe2015** [comment]
> We need a list of nixos friendly cloud providers! :-)

https://nixos.wiki/wiki/NixOS_friendly_hosters

**Comment by nixinator** [comment]
needs to be on nixos.org :-) for more visibility.

**Comment by wmertens** [comment]
Woah that kexec tool (https://github.com/cleverca22/nix-tests/tree/master/kexec) is a good idea! Just run a binary and your system is suddenly a NixOS installer on a ramdisk.

**Comment by nixinator** [comment]
> Woah that kexec tool (https://github.com/cleverca22/nix-tests/tree/master/kexec) is a good idea! Just run a binary and your system is suddenly a NixOS installer on a ramdisk.

MIND BLOWN

**Comment by AkechiShiro** [comment]
How is this different from NixOS lustrate ? 
https://nixos.org/manual/nixos/stable/#sec-installing-from-other-distro

**Comment by irisjae** [comment]
I've been told that https://github.com/nix-community/nixos-anywhere can be preferred over nixos-infect. How does nixos-anywhere compare to lustrate?

**Comment by aszlig** [comment]
> How is this different from NixOS lustrate ?

The difference is mainly that it would repartition whilst keeping your data, but given that I haven't had time to continue working on this project, I'm now officially marking this as dead since there are other options and stuff such as [disko](https://github.com/nix-community/disko) for declarative partitioning.
